test-suite/exonum-py-tests/exonum_tests/migrate.py

```diff
@@ -0,0 +1,280 @@
+import time
+import unittest
+
+from exonum_client import ExonumClient
+from exonum_launcher.action_result import ActionResult
+from exonum_launcher.configuration import Configuration
+from exonum_launcher.explorer import NotCommittedError
+from exonum_launcher.launcher import Launcher
+
+from suite import (
+    assert_processes_exited_successfully,
+    run_4_nodes,
+    wait_network_to_start,
+    generate_config,
+    generate_migration_config,
+)
+
+INSTANCE_NAME = "cryptocurrency"
+
+
+class MigrationTests(unittest.TestCase):
+    """Tests for a checking service migration mechanism."""
+
+    def setUp(self):
+        self.network = run_4_nodes("cryptocurrency-migration")
+        self.addCleanup(self._tear_down, False)
+        wait_network_to_start(self.network)
+
+    def wait_for_api_restart(self):
+        """Waits until the API servers of nodes are restarted after the set
+        of active services has changed."""
+
+        time.sleep(0.25)
+        wait_network_to_start(self.network)
+
+    def test_full_migration_flow(self):
+        """Tests full service migration flow."""
+
+        host, public_port, private_port = self.network.api_address(0)
+        client = ExonumClient(host, public_port, private_port)
+
+        # Deploy a service with 0.2.0 version.
+        instances = {INSTANCE_NAME: {"artifact": "cryptocurrency"}}
+        config_dict = generate_config(self.network, instances=instances)
+        deploy_config = Configuration(config_dict)
+
+        with Launcher(deploy_config) as launcher:
+            launcher.deploy_all()
+            launcher.wait_for_deploy()
+
+            self.wait_for_api_restart()
+            explorer = launcher.explorer()
+            for artifact in launcher.launch_state.completed_deployments():
+                deployed = explorer.is_deployed(artifact)
+                self.assertEqual(deployed, True)
+
+        # Stop the working service with version 0.1.0.
+        instances = {INSTANCE_NAME: {"artifact": "cryptocurrency", "action": "stop"}}
```

Comment:
Could also make sense to create a test, where the service is frozen instead of stopped.